platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/CheckedFingerprint.kt

```diff
@@ -28,8 +28,27 @@ sealed class CheckedFingerprint {
     object Valid : CheckedFingerprint()
 
     // The entry cannot be reused at all and should be recreated from scratch
-    class EntryInvalid(val reason: StructuredMessage) : CheckedFingerprint()
+    class EntryInvalid(val buildPath: Path, val reason: StructuredMessage) : CheckedFingerprint()
 
     // The entry can be reused, however the values for certain projects cannot be reused and should be recreated
-    class ProjectsInvalid(val reason: StructuredMessage, val invalidProjects: Set<Path>) : CheckedFingerprint()
+    class ProjectsInvalid(
+        /**
+         * First project for which an invalidation was detected.
+         */
+        val firstInvalidated: Path,
```

Comment:
*continued bikeshedding warning*

I was thinking that there is no strong motivation to tie the invalidation reason to the first-invalidated-project here. In the future, we might decide to use a summary as a reason (perhaps, listing all or some projects instead of just one).

Given this, we don't need to store the key for the map, but we can store the reason itself. It would be the caller's responsibility to decide what reason that should be.